### PR TITLE
Welcome telemetry footer

### PR DESCRIPTION
**Pull request by digitarald**
After going back and forth, here is my kinda-working-but-slightly-broken attempt.

I added a few `FIXME` annotations with thoughts and open items:

- [x] Properly render the telemetry notice
- [x] Ensure rock-solid logic for ensuring Welcome page opening on first-time opening
- [ ] Keep `telemetryOptOutUrl` product config or migrate to `telemetryOptOut` boolean
- [x] Align logic for showing the footer for logic that ensures Welcome page first-time
- [x] Remove `contrib/welcome/telemetryOptOut/*` if we are happy with this logic

I'd love to hand off the remaining work that needs an experienced dev to get the logic and details right. Thank you for offering, @isidorn (I think you did, but correct me if I am wrong)!

### CODE REVIEW

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts [hunk 1]
```diff
@@ -1244,7 +1246,7 @@ export class GettingStartedPage extends EditorPane {
 		super.clearInput();
 	}
 
-	private buildCategorySlide(categoryID: string, selectedStep?: string) {
+	private buildCategorySlide(categoryID: string, selectedStep?: string, telemetryFooter?: boolean) {
```
Comment (isidorn) [inline src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:1249]: You are not using the `telemetryFooter` flag anywhere.

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts [hunk 2]
```diff
@@ -1317,7 +1319,12 @@ export class GettingStartedPage extends EditorPane {
 		this.detailsScrollbar = this._register(new DomScrollableElement(stepsContainer, { className: 'steps-container' }));
 		const stepListComponent = this.detailsScrollbar.getDomNode();
 
-		reset(this.stepsContent, categoryDescriptorComponent, stepListComponent, this.stepMediaComponent);
+		const categoryFooter = $('.getting-started-footer');
+		// FIXME: This rendering doesn't actually work and is probably not right anwyays.
+		categoryFooter.append(renderFormattedText(localize('telemetryFooter',
```
Comment (isidorn) [inline src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:1324]: As you say rendering is broken, you need to render proper markdown.
This is what I currently see.

Screenshot 2021-08-12 at 11 55 16

File: src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts [hunk 3]
```diff
@@ -47,20 +52,29 @@ export class WelcomePageContribution implements IWorkbenchContribution {
 			const hasBackups = await this.workingCopyBackupService.hasBackups();
 			if (hasBackups) { return; }
 
-			// Open the welcome even if we opened a set of default editors
-			if (!this.editorService.activeEditor || this.layoutService.openedDefaultEditors) {
-				const startupEditorSetting = this.configurationService.inspect<string>(configurationKey);
+			// Ensure Welcome page for first-launch, no matter what is open.
+			// FIXME: Logic inspired by contrib/welcome/telemetryOptOut/browser/telemetryOptOut.ts.
+			// FIXME: Migrate telemetryOptOutUrl to boolean telemetryOptOut
+			if (this.productService.telemetryOptOutUrl && !this.storageService.get(telemetryOptOutStorageKey, StorageScope.GLOBAL)) {
```
Comment (isidorn) [inline src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts:58]: If I immediately went to the welcome page from the getting started I would not get the telemetry footer.
I found this surprising, as I think we should show this until the user clicks on it at least?

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@digitarald cool stuff! I like how this looks in the getting started view. There is work needed to polish this and make this nice, I have commented in the PR with some suggestions and some issues which I found. 
I would also love if @JacksonKearl gives this a review since this touches the `Welcome` and the `Getting Started`.

@digitarald once you do a bit more polish and address some of our comments let me know if you get stuck and then I can help with coding to wrap this up.

I would also suggest to simply get rid off `contrib/welcome/telemetryOptOut` as part of this PR as you mentioned in your initial commit.

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts [hunk 4]
```diff
@@ -893,7 +893,6 @@ export class GettingStartedPage extends EditorPane {
 
 		const someStepsComplete = this.gettingStartedCategories.some(categry => categry.content.type === 'steps' && categry.content.stepsComplete);
 		if (!someStepsComplete && !this.hasScrolledToFirstCategory) {
```
[the remark this answers is not in the document]
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:895]: Just to clarify, do you mean like 896 below this comment?

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts (hunk 2 above)
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:1324]: Do you any recommendation on how to fix it, like which specific component to use for this kind of simpler paragraph with command-firing links?

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.css [hunk 5]
```diff
@@ -473,20 +473,21 @@
...
 	display: grid;
 	grid-template-columns: 1fr 5fr 1fr 7fr 1fr;
-	grid-template-rows: calc(25% - 100px) 100px max-content 1fr;
+	grid-template-rows: calc(25% - 100px) 100px max-content 1fr 1fr;
 	grid-template-areas:
-		". back  . media ."
-		". title . media ."
-		". steps . media ."
-		". .     . media .";
+		". back   .      media  ."
```
[the remark this answers is not in the document]
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.css:478]: @JacksonKearl I added the formatting to keep the tabular alignment. Wdyt?

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts (hunk 1 above)
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:1249]: Right, my bad. Was just for testing to make sure it always shows up.

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts (hunk 2 above)
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:1324]: Fixed using MarkdownRenderer.

File: src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts (hunk 3 above)
Comment (digitarald) [reply src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts:58]: The goal is to make this more of an informative notice that should be visible on first open but doesn't require interaction (in contrast to the current sticky notification).

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts [hunk 6]
```diff
@@ -904,7 +903,10 @@ export class GettingStartedPage extends EditorPane {
 				if (first) {
 					this.currentCategory = first;
 					this.editorInput.selectedCategory = this.currentCategory?.id;
-					this.buildCategorySlide(this.editorInput.selectedCategory);
+					// FIXME: Clarify if this should be in buildCategorySlide.
```
Comment (digitarald) [inline src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:906]: @JacksonKearl your input on how this state should be carried over to the UI would be greatly appreciated.

File: src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (isidorn) [reply src/vs/workbench/contrib/welcome/gettingStarted/browser/gettingStarted.ts:896]: Yeah

File: src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts [hunk 7]
```diff
@@ -36,7 +39,9 @@ export class WelcomePageContribution implements IWorkbenchContribution {
 		@ILifecycleService private readonly lifecycleService: ILifecycleService,
 		@IWorkbenchLayoutService private readonly layoutService: IWorkbenchLayoutService,
 		@ICommandService private readonly commandService: ICommandService,
-		@IWorkbenchEnvironmentService private readonly environmentService: IWorkbenchEnvironmentService
+		@IWorkbenchEnvironmentService private readonly environmentService: IWorkbenchEnvironmentService,
+		@IProductService private readonly productService: IProductService,
```
Comment (isidorn) [inline src/vs/workbench/contrib/welcome/page/browser/welcomePage.ts:43]: `ProductService` does not seem to be used anywhere. Are you sure you need it?

**Comment by isidorn** [comment]
Let me try to answer some questions from the initial comment:

1. You are using the `MarkdownRenderer` in a correct way it seems. This needs some debugging, which I can do, or maybe @JacksonKearl has ideas
2. I think that we either open the Welcome Page or the Getting Started page on first open. Though you would know better what exact experiments we are running. I also think we have some settings so users control what gets opened, but that is not the default.
3. I like keeping PRs small, so for now I suggest that we simply keep `telemetryOptOutUrl` in the product config. This PR should only change the visuals, not how it all works in the background.
4. Makes sense. We should always show, until users clicks?
5. Yes, please remove `contrib/welcome/telemetryOptOut/* `. We love removing code 🎉

**Comment by digitarald** [comment]
> I like keeping PRs small, so for now I suggest that we simply keep telemetryOptOutUrl in the product config. This PR should only change the visuals (when and how we show), not how it all works in the background.

If I remove the `telemetryOptOut` contrib, it is the only place that `telemetryOptOutUrl` was used. Should I revert to use `telemetryOptOutUrl` in my new code as well, vs checking if telemetry is enabled?

> Makes sense. We should always show, until users clicks?
Also discussed in https://github.com/microsoft/vscode/pull/130641#discussion_r687572605, I wanted to explicitly avoid requiring an interaction and be ephemeral. The telemetry setting is now also exposed in the Settings menu to make it easier to find.

**Comment by JacksonKearl** [comment]
Thanks for this! I can take over, I have some outstanding changes to getting started that I'll need to merge this into anyways, should be in by EOD

**Comment by digitarald** [comment]
Thank you @JacksonKearl . I did push a few more fixes from the feedback. Let me know if there are open questions on heuristics and behavior.

**Comment by JacksonKearl** [comment]
Merged in the recent changes I made to getting started and hardened some of the logic for showing the notice when it hasn't been shown before. Also moved the logic for controlling whether the notice is shown into the input which helps clean things up a bit. 

Wondering if it might make sense to have the notice at a fixed position at the bottom of the screen regardless of the slide showing? So, if the telemetry prompt hasn't been shown before then keep it at the bottom of the welcome page until the tab is closed. 

The way I have it implemented currently is that it shows up on all details pages, but only for that first welcome page that loads when there's no telemetry opt-out shown flag set.

**Comment by digitarald** [comment]
> Wondering if it might make sense to have the notice at a fixed position at the bottom of the screen regardless of the slide showing? So, if the telemetry prompt hasn't been shown before then keep it at the bottom of the welcome page until the tab is closed.

Hiding it when moving between walkthroughs does free up extra vertical space that is much needed on smaller viewports. The users that are aware of telemetry and want to change it are also served by the newly added settings menu to make it easier to jump directly to the settings section. So with the direct link we added in the footer, this is an overall win for users who want more control over data collection.

Given the above, the current patch is a solid start and we can adjust based on feedback.

**Comment by isidorn** [comment]
@digitarald maybe create a small test plan item for this?
In case you do feel free to assign it to me so I give it a thorough test pass next week.

**Pull request merged**